yoga/node/Node.h

```diff
@@ -31,6 +31,128 @@ namespace facebook::yoga {
+          : node_(node),
+            childIndex_(childIndex),
+            backtrack_(std::move(backtrack)) {}
+
+      Node* operator*() const {
+        return node_->getChild(childIndex_);
+      }
+
+      Iterator& operator++() {
+        next();
+        currentNodeIndex_++;
+        return *this;
+      }
+
+      Iterator operator++(int) {
+        Iterator tmp = *this;
+        ++(*this);
+        return tmp;
+      }
+
+      size_t index() const {
+        return currentNodeIndex_;
+      }
+
+      friend bool operator==(const Iterator& a, const Iterator& b) {
+        return a.node_ == b.node_ && a.childIndex_ == b.childIndex_;
+      };
+
+      friend bool operator!=(const Iterator& a, const Iterator& b) {
+        return a.node_ != b.node_ || a.childIndex_ != b.childIndex_;
+      };
+
+     private:
+      void next() {
+        if (childIndex_ + 1 >= node_->getChildCount()) {
+          // if the current node has no more children, try to backtrack and
+          // visit its successor
+          if (backtrack_.empty()) {
+            // if there are no nodes to backtrack to, the last node has been
+            // visited
+            node_ = nullptr;
+            childIndex_ = -1;
+          } else {
+            // pop and restore the latest backtrack entry
+            const auto back = backtrack_.back();
+            backtrack_.pop_back();
+            node_ = back.first;
+            childIndex_ = back.second;
+
+            // go to the next node
+            next();
+          }
+        } else {
+          // current node has more children to visit, go to next
+          ++childIndex_;
+          // skip all display: contents nodes, possibly going deeper into the
+          // tree
+          skipContentsNodes();
+        }
+      }
+
+      void skipContentsNodes() {
+        // get the node that would be returned from the iterator
+        auto currentNode = node_->getChild(childIndex_);
+        while (currentNode->style().display() == Display::Contents &&
+               currentNode->getChildCount() > 0) {
+          // if it has display: contents set, it shouldn't be returned but its
+          // children should in its place push the current node and child index
+          // so that the current state can be restored when backtracking
+          backtrack_.push_back({node_, childIndex_});
+          // traverse the child
+          node_ = currentNode;
+          childIndex_ = 0;
+
+          // repeat until a node without display: contents is found in the
+          // subtree or a leaf is reached
+          currentNode = currentNode->getChild(childIndex_);
+        }
+
+        // if no node without display: contents was found, try to backtrack
+        if (currentNode->style().display() == Display::Contents) {
+          next();
+        }
+      }
+
+      const Node* node_;
+      size_t childIndex_;
+      size_t currentNodeIndex_{0};
+      Backtrack backtrack_;
```

Comment:
Most of the time it will be an empty vector, so no heap allocation, but I wonder if this could refer to e.g. single parent iterator (and that iterator could point to parent), such that iteration only allocates memory on the stack.

Reply:
I don't think this is easily achievable. I see two main problems with it:
- incrementing the iterator would have to return a new instance with a pointer to the previous one when going deeper into the tree, we would need to prolong the lifetime of the previous iterator somehow, so that it's valid when actually backtracking.
- modifying the "parent" iterator would possibly break backtracking logic. To prevent that, each iterator would need to keep a deep copy of all ancestors.